fs: nvs: fix invalid block compare when data CRC is enabled

yolaval opened the pull request:
When `nvs_write` is called, the `nvs_flash_block_cmp` is used to check if the new data to be written matches the data already on flash (flash write is skipped and 0 is returned if matching). This check always fail when `CONFIG_NVS_DATA_CRC` is enabled, caused by the `NVS_DATA_CRC_SIZE` being added to the `len` parameter. The pointer to the new data passed to the compare function does not already have the CRC part added, while the data on flash does and the size to be compared includes the CRC section. By removing the addition of `NVS_DATA_CRC_SIZE` to the compare size, only the data without CRC is compared, which will make the compare work in both cases.

Tested locally on my NUCLEO-G491RE board.

de-nordic commented:
@yolaval I think this should be considered a bug. Would it be possible for you to log Zephyr issue for this and mark this PR as a fix?

RICCIARDI-Adrien commented:
Indeed, the data CRC is added in `nvs_flash_data_wrt()`, which is called after `nvs_flash_block_cmp()` has been called.
Good catch @yolaval, and thanks for the fix !

RICCIARDI-Adrien commented on subsys/fs/nvs/nvs.c:
```diff
@@ -1136,7 +1136,7 @@ ssize_t nvs_write(struct nvs_fs *fs, uint16_t id, const void *data, size_t len)
 		} else if (len + NVS_DATA_CRC_SIZE == wlk_ate.len) {
 			/* do not try to compare if lengths are not equal */
 			/* compare the data and if equal return 0 */
```
Nit : I would suggest to add a little comment here, like "Data CRC is not taken into account because it is not yet appended to the data buffer".

yolaval replied:
Done.

Commit:
fs: nvs: fix invalid block compare when data CRC is enabled

When nvs_write is called, the nvs_flash_block_cmp is used to check if
the new data to be written matches the data already on flash. This check
always fail when CONFIG_NVS_DATA_CRC is enabled, caused by the
NVS_DATA_CRC_SIZE being added to the len parameter. The pointer to the
new data does not already have the CRC part added, while the data on
flash does, and the size to be compared includes CRC section.
By removing the addition of NVS_DATA_CRC_SIZE to the compare size, only
the data without CRC is compared, which will make the compare work in
both cases.

Signed-off-by: Yonas Alizadeh <yonas.alizadeh@alfalaval.com>

yolaval commented:
@Laczen sorry for new force push after approval. Had to fix a trailing whitespace that had sneaked in.

de-nordic commented:
> Indeed, the data CRC is added in `nvs_flash_data_wrt()`, which is called after `nvs_flash_block_cmp()` has been called. Good catch @yolaval, and thanks for the fix !

Should there be issue logged in Zephyr for that?
